File: ArchPropagation/AP_MATLAB/synthetic.m
clc; clear;
format compact
tic %note! Takes roughly 6 minutes to run
% Number of components in the system
nodes=1024;
% Erdos-Renyi connection probability
% p=[.9 .8 .7 .6 .5 .4 .3 .2 .1];
% Modular Perturbation Probability
p = [0.05 0.025 0.01 0.005 0.005 0.001];%p = [0.05 0.025 0.01 0.005 0.025 0.001];
% Number of eigenvectors to retain
k=100;
% Number of computation loops
g=100;
% Memory initialization
Pt=cell(1,length(p));
a=cell(length(p),g);
x=cell(length(p),g);
y=cell(length(p),g);
am=zeros(length(p),k);
xm=zeros(length(p),k);
ym=zeros(length(p),k);
% Reference system
modules=8;

% Ideal Modular
% b=modular(nodes,modules,1,nodes/modules,0,1);

% Modular
% b=modular(nodes,modules,.9,nodes/modules,.6,1);


% Hierarchically Modular
% Number of levels of hierarchy (should be >= 2)
% levels = 4;
b=modular(nodes,modules,.9,nodes/modules,.5,.5);

% Antenna case study
% load('antennacpm');
% Convert to a component DSM
% b=directlikelihood;
% b(b>0)=1;
b=triu(b);


b=b+b';

b(logical(eye(size(b))))=0;

%%
for i=1:length(p),
    % Compute perturbation
    for j=1:g,
        % Random Peturbation Matrix
%         Er1=PerturbMatrix(b,p(i),.5,.5);
%         Er2=PerturbMatrix(b,p(i),.9,.1);
%         Er3=PerturbMatrix(b,p(i),.1,.9);
%         Pt{i}=Er1;
        %            Modular Perturbation Matrix
        %         Er1=ModularPerturbMatrix(b,modules,0,p(i),.5,.5);
        %         Er2=ModularPerturbMatrix(b,modules,0,p(i),.9,.1);
        %         Er3=ModularPerturbMatrix(b,modules,0,p(i),.1,.9);
        %         Pt{i}=Er1;
        % Hierarchically Modular Perturbation Matrix
        Er1=HMPerturbMatrix(b,nodes,modules,0,p(i),.2,.8);
        Er2=HMPerturbMatrix(b,nodes,modules,0,p(i),.3,.7);
        Er3=HMPerturbMatrix(b,nodes,modules,0,p(i),.4,.6);
        a{i,j}=ComputeSine(b,Er1,k);
        x{i,j}=ComputeSine(b,Er2,k);
        y{i,j}=ComputeSine(b,Er3,k);
    end
    % Erdos-Renyi random perturbation
    %r=erdosRenyi(nodes,p(i),4);
    %Er=full(r.Adj);
    %Er=triu(Er);
    %Er=Er+Er';
    %Er(logical(eye(size(Er))))=0;
    %Pt{i}=Er;
    % Compute perturbation
    %a(i,:)=ComputeSine(b,Er,k);
    % Somwrita's random perturbation
    %      H=RandNetwork(nodes,p(i));
    %      Er=ones(nodes,nodes);
    %      Er(H==0)=-1;
    %      Er(logical(eye(size(Er))))=0;
    %      Er = sparse(Er);
    %      Pt{i}=Er;
end
% Plot the results
h1=figure('Name','Eigenvector Rotation');
subplot(2,2,1);
%scatter(1:size(brot,2),sort(real(brot),'descend'),'*');
imshow(b);
%xlabel('Eigenvector index');
%ylabel('Angle of rotation');
title('Unperturbed System');
%subplot(2,2,2);
%imshow(b+Pt{1});
%title('Perturbed System p=.9');
%markers = ['o','+','.','x','s','d','^','p','h'];
%h2=figure('Name','Perturbed System');
%hold on;
%for i=1:length(p),
%    scatter(1:size(a(i,:),2),sort(real(a(i,:)),'descend'),markers(i));
%end

%Find the mean perturbations
%am=mean(a,1);
% This calculates the mean perturbation per value of p
for i=1:length(p),
    am(i,:)=mean(cell2mat(a(i,:)'),1);
    xm(i,:)=mean(cell2mat(x(i,:)'),1);
    ym(i,:)=mean(cell2mat(y(i,:)'),1);
end

agrandmean=mean(am,1);
xgrandmean=mean(xm,1);
ygrandmean=mean(ym,1);

subplot(2,2,2);
%scatter(1:size(am,2),sort(real(am),'descend'),'Marker','o','MarkerEdgeColor','red');
scatter(1:size(agrandmean,2),agrandmean,'Marker','o','MarkerEdgeColor','red');
%scatter(1:size(a(i,:),2),sort(real(a(i,:)),'descend'),'o');
xlabel('Eigenvector index');
ylabel('Angle of rotation');
%legend('p=.9','p=.8','p=.7','p=.6','p=.5','p=.4','p=.3','p=.2','p=.1');
title('Perturbed pw=.5 pu=.5');
subplot(2,2,3);
scatter(1:size(xgrandmean,2),xgrandmean,'Marker','o','MarkerEdgeColor','red');
xlabel('Eigenvector index');
ylabel('Angle of rotation');
title('Perturbed pw=.9 pu=.1');
subplot(2,2,4);
scatter(1:size(ygrandmean,2),ygrandmean,'Marker','o','MarkerEdgeColor','red');
xlabel('Eigenvector index');
ylabel('Angle of rotation');
title('Perturbed pw=.1 pu=.9');



toc
% Print Results
str=sprintf('%u & %.1f & %.1f & %.1f & %u & %.1f & %.1f & %.1f & %u & %.1f & %.1f & %.1f & %.1f',nace, ...
    mace, mabe, mabe-mace, nxce, mxce, mxbe, mxbe-mxce, nyce, myce, mybe, mybe-myce, max([mabe-mace,mxbe-mxce,mybe-myce])-min([mabe-mace,mxbe-mxce,mybe-myce]));
fprintf(str);

File: ArchPropagation/AP_Python/synthetic.py
# -*- coding: utf-8 -*-
"""
Created on Fri Jun 23 09:28:15 2023

@author: colej

synthetic.py: Python translation of synthetic.m written by ___

Note: this is computationally intensive due to translation from MATLAB,
the loops for large sample size, and my own limited skills as a programmer
"""
# import required packages and functions from support file
import numpy as np
import pandas as pd
from arch_prop_functions import *
import matplotlib.pyplot as plt
import time

beginning = time.time()
# reset random number generator for reproducability (same as MATLAB)
np.random.seed(1337)

# Number of components in the system
nodes = 256
# modular perturbation probability
# p = np.array([0.05, 0.025, 0.01, 0.005, 0.025, 0.001])
p = np.array([0.05, 0.025])#, 0.01, 0.005, 0.025, 0.001])
# Number of eigenvectors to retain
k = 100
# Number of computation loops
g = 100


# %%
# memory initialization
Pt = pd.DataFrame(index=range(1), columns=range(len(p)))
a = pd.DataFrame(index=range(len(p)), columns=range(g))
x = pd.DataFrame(index=range(len(p)), columns=range(g))
y = pd.DataFrame(index=range(len(p)), columns=range(g))
am = pd.DataFrame(np.zeros((len(p), k)))
xm = pd.DataFrame(np.zeros((len(p), k)))
ym = pd.DataFrame(np.zeros((len(p), k)))

# %%reference system
modules = 8

b = modular(nodes, modules, 0.9, int(nodes / modules), 0.5, 0.5)
b = np.triu(b.to_numpy())
b = (
    b + b.T
)  # +5*np.identity(nodes) #add values to diagonal to filter for removal

b[
    (np.arange(0, nodes), np.arange(0, nodes))
] = 0  # turn diagonal values into zeros (this should be simpler to do but idk)

# plt.imshow(b)
# plt.show()


#%% Doing the perturbations\

for i in range(0, len(p)):
    print("Working on Perturbation Probability", p[i])
    for j in range(0, g):
        Er1 = HMPerturbMatrix(b, nodes, modules, 0, p[i], 0.5, 0.5)
        Er2 = HMPerturbMatrix(b, nodes, modules, 0, p[i], 0.9, 0.1)
        Er3 = HMPerturbMatrix(b, nodes, modules, 0, p[i], 0.1, 0.9)
        a.iloc[i, j] = ComputeSine(b, Er1, k)
        x.iloc[i, j] = ComputeSine(b, Er2, k)
        y.iloc[i, j] = ComputeSine(b, Er3, k)

print("Calculations elapsed in ", time.time() - beginning, " seconds")

#%% Calculating mean perturbations per value of p
# a.iloc[0,:] = np.arange(0,100)
a = a.to_numpy()
x = x.to_numpy()
y = y.to_numpy()

# I don't think there's any point to this code because it's not used elsewhere but is in the original
for i in range(0, len(p)):
    am.iloc[i, :] = np.mean(a[i, :])
    xm.iloc[i, :] = np.mean(x[i, :])
    ym.iloc[i, :] = np.mean(y[i, :])


agrandmean = np.mean(a)
xgrandmean = np.mean(x)
ygrandmean = np.mean(y)

#%% Plotting the results
figs, axs = plt.subplots(2, 2)
axs[0, 0].imshow(b)
axs[0, 0].set_xticks([])
axs[0, 0].set_yticks([])
axs[0, 0].set_title("Unperturbed System")

axs[0, 1].scatter(np.arange(0, k), agrandmean, marker=".")
axs[0, 1].set_title("Perturbed pw=0.5, pu=0.5")

axs[1, 0].scatter(np.arange(0, k), xgrandmean, marker=".")
axs[1, 0].set_title("Perturbed pw=0.9, pu=0.1")

axs[1, 1].scatter(np.arange(0, k), ygrandmean, marker=".")
axs[1, 1].set_title("Perturbed pw=0.1, pu=0.9")

#%% Calculating and printing eigenvalues

# Number of eigenvalues in community eigenspace
nace = RotationPairs(agrandmean)
# Number of eigenvalues in bulk eigenspace
nabe = k - nace
# Mean angle community eigenspace
mace = np.mean(agrandmean[0:nace])
# Mean angle bulk eigenspace
mabe = np.mean(agrandmean[nace + 1 : k])

# Number of eigenvalues in community eigenspace
nxce = RotationPairs(xgrandmean)
# Number of eigenvalues in bulk eigenspace
nxbe = k - nxce
# Mean angle community eigenspace
mxce = np.mean(xgrandmean[0:nxce])
# Mean angle bulk eigenspace
mxbe = np.mean(xgrandmean[nxce + 1 : k])

# Number of eigenvalues in community eigenspace
nyce = RotationPairs(ygrandmean)
# Number of eigenvalues in bulk eigenspace
nybe = k - nyce
# Mean angle community eigenspace
myce = np.mean(ygrandmean[0:nyce])
# Mean angle bulk eigenspace
mybe = np.mean(ygrandmean[nyce + 1 : k])


#'%u & %.1f & %.1f & %.1f & %u & %.1f & %.1f & %.1f & %u & %.1f & %.1f & %.1f & %.1f',nace, ...
# mace, mabe, mabe-mace, nxce, mxce, mxbe, mxbe-mxce, nyce, myce, mybe, mybe-myce,
# max([mabe-mace,mxbe-mxce,mybe-myce])-min([mabe-mace,mxbe-mxce,mybe-myce]));
print(
    "%d" % nace,
    "& %0.1f" % mace,
    "& %0.1f" % mabe,
    "& %0.1f" % (mabe - mace),
    "& %d" % nxce,
    "& %0.1f" % mxce,
    "& %0.1f" % mxbe,
    "& %0.1f" % (mxbe - mxce),
    "& %d" % nyce,
    "& %0.1f" % myce,
    "& %0.1f" % mybe,
    "& %0.1f" % (mybe - myce),
    "& %0.1f"
    % (
        max(mabe - mace, mxbe - mxce, mybe - myce)
        - min(mabe - mace, mxbe - mxce, mybe - myce)
    ),
)
